src/utils/helpers/GraphHelpers.ts

```diff
@@ -0,0 +1,184 @@
+
+    if (isConsumer && isProvider) {
+        return 'M4.5,35l9.37,14.59L4.5,64.18h' + (labelTextWidth + 45) + 'l9-14.59L' + (labelTextWidth + 49.5) + ',35H4.5z';
+    }
+
+    if (isProvider) {
+        return 'M' + (labelTextWidth + 49.5) + ',35H4.5l9.37,14.59L4.5,64.18h' + (labelTextWidth + 45);
+    }
+
+    if (isConsumer) {
+        return 'M4.5,64.18h' + (labelTextWidth + 45) + 'l9.42-14.59L' + (labelTextWidth + 49.5) + ',35H4.5';
+    }
+
+    return 'M4.5,64.18h' + (labelTextWidth + 55) + 'L' + (labelTextWidth + 59.5) + ',35H4.5';
+}
+
+export function handleDrag(simulation: Simulation<DependencyNode, DependencyLink>) {
+    function dragStarted(node: DependencyNode) {
+        if (!d3.event.active) {
+            simulation.alphaTarget(0.3).restart();
+        }
+        node.fx = node.x;
+        node.fy = node.y;
+    }
+
+    function dragged(node: DependencyNode) {
+        node.fx = d3.event.x;
+        node.fy = d3.event.y;
+    }
+
+    function dragEnded(node: DependencyNode) {
+        if (!d3.event.active) {
+            simulation.alphaTarget(0);
+        }
+        node.fx = null;
+        node.fy = null;
+    }
+
+    return d3
+        .drag<SVGGElement, DependencyNode>()
+        .on('start', dragStarted)
+        .on('drag', dragged)
+        .on('end', dragEnded);
+}
+
+export function createSimulation(nodes: DependencyNode[], links: DependencyLink[], width: number, height: number) {
+    return d3
+        .forceSimulation(nodes)
+        .force(
+            'dependency',
+            d3
+                .forceLink<DependencyNode, DependencyLink>(links)
+                .distance(180)
+                .id((node: DependencyNode) => node.name)
+        )
+        .force('center', d3.forceCenter(width / 2, height / 2))
+        .force('y', d3.forceY(0.5))
+        .force('collide', d3.forceCollide(140))
+        .force('nodeCollide', d3.forceCollide(140));
+}
+
+export function createSVGContainer(
+    container: HTMLDivElement,
+    width: number,
+    height: number
+): d3.Selection<SVGSVGElement, DependencyNode, Element, HTMLElement> {
+    return d3
+        .select<Element, DependencyNode>(`#${container.id}`)
+        .append('svg')
+        .attr('preserveAspectRatio', 'xMidYMid meet')
+        .attr('viewBox', `0 0 ${width} ${height}`)
+        .attr('width', width)
+        .attr('height', height);
+}
+
+export function createZoom(svgContainer: GraphContainer): d3.Selection<SVGGElement, DependencyNode, Element, HTMLElement> {
+    const zoomLayer = svgContainer.append('g');
+
+    const zoomed = () => {
+        zoomLayer.attr('transform', d3.event.transform);
+    };
+
+    svgContainer
+        .call(
+            d3
+                .zoom<SVGSVGElement, DependencyNode>()
+                .scaleExtent([1 / 2, 12])
+                .on('zoom', zoomed)
+        )
+        .on('dblclick.zoom', null);
+
+    return zoomLayer;
+}
+
+export function getLabelTextHeight(this: Node, dim: 'width' | 'height') {
```

Comment:
Nice.
I didn't know that we can use ``this`` word as parameter and then call for example ``getLabelTextHeight.call('width');``
👍 

Could we write ? : 
```typescript
export function getLabelTextHeight(someNode: Node, dim: 'width' | 'height') {
```
and then for example:
```typescript
getLabelTextHeight.(this, 'width');
```
instead ``getLabelTextHeight.call(this, 'width')`` ?

Reply:
Done

Reply:
We probably should consult more experienced developers, but for me personally, passing around `this` always seems a little bit risky. Since objects are passed by reference, you can incidentally mutate the object being passed as the function argument without being 100% conscious of it. But maybe my concerns are out of place and our senior developers will clarify this matter.

Reply:
I had to choose between passing `this` and finding it through iterating once again on all nodes. Thanks for pointing this out, it should be taken into account.